test/wadler_test.rb

```diff
@@ -0,0 +1,155 @@
+# frozen_string_literal: true
+
+require 'prettyprint'
+
+require_relative 'lib'
+
+def test_lib_with_ref(margin, expected, builder_block)
+  _(builder_block.call(margin)).must_equal expected
+  _(builder_block.call(margin, false)).must_equal expected
+end
+
+describe 'Wadler tests' do
+  it 'must work like ruby\'s PrettyPrint library' do
+    builder_block = proc { |margin, is_our_lib = true|
+      out = if is_our_lib
+        Oppen::Wadler.new(margin:)
+      else
+        PrettyPrint.new(''.dup, margin)
+      end
+      out.group {
+        out.group {
+          out.group {
+            out.group {
+              out.text 'hello'
+              out.breakable
+              out.text 'a'
+            }
+            out.breakable
+            out.text 'b'
+          }
+          out.breakable
+          out.text 'c'
+        }
+        out.breakable
+        out.text 'd'
+      }
+      if !is_our_lib
+        out.flush
+      end
+      out.output
+    }
+
+    expected = <<~LANG.chomp
+      hello
+      a
+      b
+      c
+      d
+    LANG
+    test_lib_with_ref(6, expected, builder_block)
+
+    expected = <<~LANG.chomp
+      hello a
+      b
+      c
+      d
+    LANG
+    test_lib_with_ref(7, expected, builder_block)
+    test_lib_with_ref(8, expected, builder_block)
+
+    expected = <<~LANG.chomp
+      hello a b
+      c
+      d
+    LANG
+    test_lib_with_ref(9, expected, builder_block)
+    test_lib_with_ref(10, expected, builder_block)
+
+    expected = <<~LANG.chomp
+      hello a b c
+      d
+    LANG
+    test_lib_with_ref(11, expected, builder_block)
+    test_lib_with_ref(12, expected, builder_block)
+
+    expected = <<~LANG.chomp
+      hello a b c d
+    LANG
+    test_lib_with_ref(13, expected, builder_block)
+  end
+
+  it 'must work with a tree class' do
+    class Tree # rubocop:disable Lint/ConstantDefinitionInBlock
+      attr_reader :string
+      attr_reader :children
+
+      def initialize(string, *children)
+        @string = string
+        @children = children
+      end
+
+      def show(out, is_our_lib)
+        out.group {
+          out.text string
+          out.nest(string.length) {
+            unless children.empty?
+              out.text '['
+              out.nest((!is_our_lib && 1) || 0) {
+                first = true
+                children.each { |t|
```

Comment:
Don't forget that we need to send the printing object from the outside, so the Tree class needs to figure out how much indentation w need in a nesting. This should be the first thing to compute before the remainder of `show`.